csrc/scheduler/matmul_utils.cpp

```diff
@@ -187,10 +195,148 @@ std::string isMatmulFusionDefinitionSupported(
+    if (!roles_map_opt.isValid()) {
+      return roles_map_opt.getErrorMsg();
+    }
+
+    const auto& roles_map = roles_map_opt.getData();
+    auto entry = roles_map.find(MatmulRole::INPUT_A);
+    std::set<TensorView*> tvs_with_roles;
+
+    if (entry != roles_map.end()) {
+      if (MATMUL_CORE_ROLES_EXPECTED_COUNT == entry->second.size()) {
+        tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+      } else {
+        return "There is more than a single fusion input that can be MMA first input";
+      }
+    } else {
+      return "No candidate in fusion inputs for MMA first input";
+    }
+
+    entry = roles_map.find(MatmulRole::INPUT_B);
+    if (entry != roles_map.end()) {
+      if (MATMUL_CORE_ROLES_EXPECTED_COUNT == entry->second.size()) {
+        tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+      } else {
+        return "There is more than a single fusion input that can be MMA second input";
+      }
+    } else {
+      return "No candidate in fusion inputs for MMA second input";
+    }
+
+    entry = roles_map.find(MatmulRole::OUTPUT_D);
+    if (entry != roles_map.end()) {
+      tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+    } else {
+      return "No candidate in fusion outputs MMA output";
+    }
+
+    // Non-core input roles are optional, no requirements for definitions
+    entry = roles_map.find(MatmulRole::INPUT_C);
+    if (entry != roles_map.end()) {
+      tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+    }
+
+    // Non-core output roles are optional, no requirements for definitions
+    entry = roles_map.find(MatmulRole::OUTPUT_AUX);
+    if (entry != roles_map.end()) {
+      tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+    }
+
+    const auto in_out_tvs_count =
+        fusion_inputs_tvs.size() + fusion_outputs_tvs.size();
+    if (in_out_tvs_count != tvs_with_roles.size()) {
+      return "Detected input/output TVs without assigned roles";
+    }
+  }
+
+  // MmaOp inputs/outputs dependencies check
+  // TODO: check to be removed when more rules are added to TV roles
+  //  calculations
+  {
+    // Check the expected path between MmaOp input and fusion inputs
+    const auto areMmaOpInputDependeciesValid = [](const Val* val) {
+      if (val->definition()->isA<BroadcastOp>()) {
+        const auto& bcast_inputs = val->definition()->inputs();
+        // BroadcastOp has single input/output, not need to check other things
+        return bcast_inputs.front()->isFusionInput() ||
+            (dynamic_cast<LoadStoreOp*>(bcast_inputs.front()->definition()) !=
+             nullptr);
+      }
+      return false;
+    };
+
+    // MmaOp input is a result of broadcast op with input being fusion input
+    for (const auto* mma_in : mma_inputs) {
+      if (!areMmaOpInputDependeciesValid(mma_in)) {
+        return "MmaOp input has unsupported dependency";
+      }
+    }
+  }
+
+  return "";
```

Comment:
Nitpick: would `std::optional<std::string>` make a better error type?

Reply:
This was a bit of existing code - and this is at multiple locations. I don't want to change it in this PR, since this may also mean modifying code that handles the return.